packages/engine/lib/error/src/frame.rs

```diff
@@ -0,0 +1,266 @@
+        // Attach E's native vtable onto the pointer to self._error
+        let unerased = (frame as *const ErrorRepr<()>).cast::<ErrorRepr<E>>();
+        // inside of vtable it's allowed to access `_error`
+        #[allow(clippy::used_underscore_binding)]
+        &(*(unerased))._error
+    }
+
+    /// Downcasts error to `target`
+    ///
+    /// # Safety
+    ///
+    /// - Layout of `Self` must match `ErrorRepr<E>`.
+    unsafe fn object_downcast<E: Context>(
+        frame: &ErrorRepr<()>,
+        target: TypeId,
+    ) -> Option<NonNull<()>> {
+        if TypeId::of::<E>() == target {
+            // Attach E's native vtable onto the pointer to self._error
+            let unerased = (frame as *const ErrorRepr<()>).cast::<ErrorRepr<E>>();
+            // inside of vtable it's allowed to access `_error`
+            #[allow(clippy::used_underscore_binding)]
+            let addr = &(*(unerased))._error as *const E as *mut ();
+            Some(NonNull::new_unchecked(addr))
+        } else {
+            None
+        }
+    }
+}
+
+#[repr(C)]
+pub struct ErrorRepr<E> {
+    vtable: &'static VTable,
+    // Must not be used directly, only through vtable
+    _error: E,
+}
+
+impl<E> ErrorRepr<E> {
+    pub fn new(error: E) -> Box<ErrorRepr<()>>
+    where
+        E: Provider + fmt::Display + fmt::Debug + Send + Sync + 'static,
+    {
+        let unerased = Box::new(Self {
+            vtable: &VTable {
+                object_drop: VTable::object_drop::<E>,
+                object_ref: VTable::object_ref::<E>,
+                object_downcast: VTable::object_downcast::<E>,
+            },
+            _error: error,
+        });
+        unsafe { Box::from_raw(Box::into_raw(unerased).cast()) }
+    }
+}
+
+#[allow(clippy::used_underscore_binding)]
+impl ErrorRepr<()> {
+    pub fn from_message<M>(message: M) -> Box<Self>
+    where
+        M: fmt::Display + fmt::Debug + Send + Sync + 'static,
+    {
+        ErrorRepr::new(WrapErr(message))
+    }
+
+    #[cfg(feature = "std")]
+    pub fn from_std<M>(error: M) -> Box<Self>
+    where
+        M: std::error::Error + Send + Sync + 'static,
+    {
+        ErrorRepr::new(StdError(error))
+    }
+
+    pub fn unerase(&self) -> &dyn Context {
+        // Use vtable to attach E's native vtable for the right original type E.
+        unsafe { (self.vtable.object_ref)(self) }
+    }
+
+    pub fn downcast<E>(&self) -> Option<NonNull<E>>
+    where
+        E: Provider + fmt::Display + fmt::Debug + Send + Sync + 'static,
+    {
+        let target = TypeId::of::<E>();
+        // Use vtable to attach E's native vtable for the right original type E.
+        let addr = unsafe { (self.vtable.object_downcast)(self, target) };
+        addr.map(NonNull::cast)
+    }
+}
+
+impl Frame {
+    /// Returns the location where this `Frame` was created.
+    #[must_use]
+    pub const fn location(&self) -> &'static Location<'static> {
```

Comment:
`Location` generation is based on `Location::caller()` which itself maps to a compiler intrinsic. As it's generating a `'static` reference, it will be generated at compile time, so the information will lie in the binary. This may bloat the file size. We should consider adding a feature flag for the location as well.

The locations are only used when an error occurs which I don't expect to be in the hot path, so the performance hit is negligible.

Reply:
Cheers for the explanation. We can add a feature flag if later analysis shows it to be a problem, no need to micro-optimise yet :)